### PR TITLE
✨  login-settings customizable per product  (🗃️ DB change)

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

✨ ``webserver.login`` settings can be now **customized** per product.  

- The settings for the ``login``  plugin are resolved by **composing** the app settings (i.e. ``app_settings.LOGIN_SETTINGS``) with the product settings (set in the postgres's ``products.login_settings`` column). Then ``app_settings.LOGIN_SETTINGS``set a "default" of base configuration where the necessary fields con be customized per product by adding overrides in ``products.login_settings``. 
- For example, let's assume two product ``X`` and ``Y`` in an app with the following login settings
```python
app_settings.LOGIN_SETTINGS = {
"LOGIN_REGISTRATION_CONFIRMATION_REQUIRED": False,
"LOGIN_REGISTRATION_INVITATION_REQUIRED": False,
"LOGIN_2FA_REQUIRED": False,
...
}
```
then product ``X`` can enforce confirmation by overriding the corresponding attribute
```python
products.login_settings = { "LOGIN_REGISTRATION_CONFIRMATION_REQUIRED": True }
```
while product ``Y`` with
```python
products.login_settings = { }
```
will not require confirmation.

###  🗃️ DB changes
  - Changes in ``products.login_settings`` defaults 
  - ⚠️ Inspect values in column ``products.login_settings`` after migration

### Dev notes
 - Note the signature of ``login.settings.get_plugin_settings(app: web. Application, product_name: str) -> LoginSettingsForProduct``
 - Note that not all compositions are possible but that is validated upon startup!
 - This approach can be extended to any other plugin settings.
 - Previous solution used a ``dataclass``  ``products.login_settings``. That choice enforced that all fields need to be set. In order to avoid that we use now ``TypedDict``.

## Extras
- ♻️ Refactors generation of passcode to utils and increases it to 6 digits as recommended (EXTRA)

## Related issue/s

- Implements part of ITISFoundation/osparc-issues#765: ``2FA env-var needs to go in the Data Base``

## How to test

- ``pytest tests/unit/**/test_product*.py`` covers changes

## Checklist

- [x] check migration against master-deploy data
- [x] ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes

### CODE REVIEW

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/10b293fdcd56_alters_product_login_settings.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    # Reassign items from two_factor_enabled -> LOGIN_2FA_REQUIRED
+    conn = op.get_bind()
+    rows = conn.execute("SELECT name, login_settings FROM products").fetchall()
+    for row in rows:
+        data = row["login_settings"] or "{}"
+        data = json.loads(data)
+        data["LOGIN_2FA_REQUIRED"] = data.pop("two_factor_enabled")
+        data = json.dumps(data)
+        conn.execute(
+            "UPDATE products SET login_settings = '{}' WHERE name = {}".format(
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/10b293fdcd56_alters_product_login_settings.py:31]: this is a migration script. no risk

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/10b293fdcd56_alters_product_login_settings.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    # Reassign items from LOGIN_2FA_REQUIRED -> two_factor_enabled
+    conn = op.get_bind()
+    rows = conn.execute("SELECT name, login_settings FROM products").fetchall()
+    for row in rows:
+        data = row["login_settings"] or "{}"
+        data = json.loads(data)
+        data["two_factor_enabled"] = data.pop("LOGIN_2FA_REQUIRED")
+        data = json.dumps(data)
+        conn.execute(
+            "UPDATE products SET login_settings = '{}' WHERE name = {}".format(
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/10b293fdcd56_alters_product_login_settings.py:55]: same as above

**Review by odeimaiz — APPROVED** [review]
👍

**Comment by odeimaiz** [comment]
Could you, please, update the frontend text "We just sent a 4-digit code to " -> "We just sent a 6 digits code to "? Merci!

**Review by sanderegg — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/login/settings.py [hunk 3]
```diff
@@ -89,6 +69,38 @@ def login_2fa_needs_sms_service(cls, v, values):
...
+    """
+    Customization of these plugin settings for a product
+
+    LoginSettings are created (and frozen) upon creation of the app and available
+    to the plugin for setup_login.
+
+    LoginSettingsForProduct is obtained composing those with the overrides defined for each
+    product in the  database.
+
+    This is initialized for each product during on_startup events at tje setup_login
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/login/settings.py:82]:
```suggestion
    This is initialized for each product during on_startup events by the setup_login
```

File: services/web/server/src/simcore_service_webserver/products_model.py [hunk 4]
```diff
@@ -71,7 +73,11 @@ class Product(BaseModel):
 
     support: Optional[list[Union[Forum, EmailFeedback, WebFeedback]]] = Field(None)
 
-    login_settings: ProductLoginSettings = Field(...)
+    login_settings: ProductLoginSettingsDict = Field(
+        ...,
+        description="Product customization of login settings. "
+        "Note that these are NOT the final plugin settings bug those are obtained from login.settings.get_plugin_settings",
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/products_model.py:79]:
```suggestion
        "Note that these are NOT the final plugin settings but those are obtained from login.settings.get_plugin_settings",
```

**Review by GitHK — COMMENTED** [review]
:+1:

**Review by GitHK — APPROVED** [review]

**Review by Surfict — APPROVED** [review]

**Comment by mrnicegyu11** [comment]
tldr
but I read the PR message, thanks a lot for this, you get my `ack` :)
